### PR TITLE
Remove ReplayingDecoder and replace its usage with ByteToMessageDecoder

### PR DESCRIPTION
**Pull request by normanmaurer**
Motivation:

There is really no need for the ReplayingDecoder at all and all its use-cases can be implemented by using the ByteToMessageDecoder directly.

Modifications:

- Replace usages of ReplayingDecoder with ByteToMessageDecoder
- Remove ReplayingDecoder and related classes
- Remove the *Compatible* marshalling decoder / encoders as their usage was disencouraged anyway and make them work without ReplayingDecoder would be costly in terms of generating exceptions etc.

Result:

Fixes https://github.com/netty/netty/issues/11659

### CODE REVIEW

File: codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5AddressDecoder.java [hunk 1]
```diff
@@ -35,16 +35,30 @@
...
+                    return null;
+                }
                 return NetUtil.intToIpAddress(in.readInt());
             }
             if (addrType == Socks5AddressType.DOMAIN) {
-                final int length = in.readUnsignedByte();
+                if (readableBytes < 1) {
+                    return null;
+                }
+                final int length = in.getUnsignedByte(in.readerIndex());
```
Comment (chrisvest) [inline codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5AddressDecoder.java:49]: I wonder if this pattern is common enough to have a `peek*` family of methods on Buffer?

**Review by chrisvest — APPROVED** [review]
Awesome

File: codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5AddressDecoder.java (hunk 1 above)
Comment (normanmaurer) [reply codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5AddressDecoder.java:49]: I think the problem would be that these would need to return "boxed" values and so would be more costly. WDYT ?

Comment (normanmaurer) [reply codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5AddressDecoder.java:49]: ah or you mean to return a value at the current index without advance the index ?

Comment (chrisvest) [reply codec-socks/src/main/java/io/netty/handler/codec/socksx/v5/Socks5AddressDecoder.java:49]: yeah, the latter

**Pull request merged**